src/server/shell/token_authority.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mir/shell/token_authority.h"
+
+#include "mir/main_loop.h"
+#include "mir/time/alarm.h"
+#include "mir/log.h"
+
+#include <algorithm>
+#include <cmath>
+#include <memory>
+#include <mutex>
+#include <uuid.h>
+
+namespace
+{
+/// Time in milliseconds that the compositor will wait before invalidating a token
+static auto constexpr timeout_ms = std::chrono::milliseconds(3000);
+}
+
+namespace msh = mir::shell;
+
+msh::TokenAuthority::Token::Token(
+    std::string token, std::optional<RevocationListener> revocation_listener) :
+    token{token},
+    revocation_listener{revocation_listener}
+{
+}
+
+msh::TokenAuthority::Token::operator std::string() const
+{
+    return token;
+}
+
+bool msh::TokenAuthority::Token::operator==(Token const& token) const
+{
+    return this->token == token.token;
+}
+
+msh::TokenAuthority::TokenAuthority(std::shared_ptr<MainLoop>&& main_loop) :
+    main_loop{std::move(main_loop)}
+{
+}
+
+auto msh::TokenAuthority::issue_token(std::optional<Token::RevocationListener> revocation_listener) -> Token
+{
+    auto token = Token{generate_token(),  revocation_listener};
+
+    auto alarm = main_loop->create_alarm(
+        [this, token]()
+        {
+            revoke_token(token);
+        });
+    alarm->reschedule_in(::timeout_ms);
+
+    {
+        std::scoped_lock lock{mutex};
+
+        revocation_alarms.emplace_back(std::move(alarm));
+        issued_tokens.insert(token);
+    }
+
+    return token;
+}
+
+auto msh::TokenAuthority::get_token_for_string(std::string const& string_token) const -> std::optional<Token>
+{
+    std::scoped_lock lock{mutex};
+
+    // Incoming string doesn't have null terminator, just like the one leaving Mir
+    auto iter = issued_tokens.find(Token{string_token, {}});
+
+    if (iter == issued_tokens.end())
+        return std::nullopt;
+
+    return *iter;
+}
+
+auto msh::TokenAuthority::is_token_valid(Token const& token) const -> bool
+{
+    std::scoped_lock lock{mutex};
+
+    // Let it pass, we'll refuse to activate when we get the token later
+    if(token == get_bogus_token())
+        return true;
+
+    return issued_tokens.contains(token);
+}
+
+auto msh::TokenAuthority::get_bogus_token() const -> Token
+{
+    return Token{"I AM A TEAPOT", {}};
```

Comment:
I'm not convinced a bogus token should have such an easily identified value. Shouldn't it be initialised identically to the non-bogus ones, but not registered in `issued_tokens`?